### ISSUE TITLE
Simulation pipeline

### ISSUE THREAD

**Issue by rvosa** [naturalis/supersmart#49]
Here's what we need to do:
* given a tree (i.e. Primates, Gentianales), replicate its general properties - birth/death rate, tree size, depth, genus size distribution - on a simulated tree (done: https://github.com/rvosa/bio-phylo/commit/b407be8390c4918abc2f0a0ee8dc2c2a0510a14e)
* given a locus presence/absence matrix, replicate its general properties - instantaneous rate matrix, linkage among columns - by simulation of a binary character
* given an alignment, replicate its general properties - substitution model, indels, length - by sequence simulation
* rarify the alignments using the presence/absence matrix
* run our two-step approach and the other megatree methods
* quantify tree-to-tree distance (Robinson-Foulds) between each of the methods and the simulated tree

**Comment by rvosa** [comment]
To do, in practical terms:
* implement a wrapper around Discrete/BayesTraits/some other likelihood-based method of binary character evolution analysis to obtain a transition matrix. Add a subclass of Bio::Phylo::Models::Substitution to store the transition matrix.
* implement a wrapper around modeltest/some other substitution model-testing program so we can pass in an alignment (maybe also a tree), get the model back and represent it using Bio::Phylo::Models::Substitution::Dna
* implement a wrapper around PhyloSim (use Statistics::R) so that we can pass in a tree, a Bio::Phylo::Models::Substitution::Dna model and requested sequence length to the wrapper and get sequences back 
* implement a wrapper around a binary character simulator

**Comment by hettling** [comment]
checking if we can use Bio::Tools::Run::Phylo::Hyphy::Modeltest

**Comment by rvosa** [comment]
Maybe better to use phangorn::modelTest - http://cran.r-project.org/web/packages/phangorn/vignettes/Trees.pdf (page 6)

**Comment by rvosa** [comment]
In any case, I will now work on the binary character if you can do the DNA?

**Comment by rvosa** [comment]
I think the idea should be that Bio::Phylo::Models::Substitution::Dna should have a ->modeltest method that invokes phangorn::modelTest to get the parameters, then instantiate the appropriate subclass (e.g. *HKY85) with those parameters. Subsequently, in Bio::Phylo::Matrices::MatrixRole, fill in the blanks for _replicate_dna analogous to what I'm doing for _replicate_binary, as of this commit: https://github.com/rvosa/bio-phylo/commit/84dcd5a85794a8dbdab2adaefc3cf917627ed950

**Comment by rvosa** [comment]
As of commit https://github.com/rvosa/bio-phylo/commit/cda7662eec322a55a1b025ac4669e7e3858dfdae simulating the binary presence/absence matrix works.

**Comment by hettling** [comment]
phangorn's modelTest functionality added with commit 1f538aef15998c1aebb3cc0c5b9e9ac605b6322a

**Comment by rvosa** [comment]
How are we doing on this?

**Comment by hettling** [comment]
At the moment `smrt-utils replicate`, which produces tree and alignments according to given input. This is almost enough for a smooth supersmart run, what needs to be done

* import artificial species into database
* import sequences into database

**Comment by hettling** [comment]
As of  a7ef4da24e3ff3eaf6b375091284f8ffea535390 we take the correct trees for replicating; the DNA substitution model is estimated with the input tree, the sequences are simulated along the replicated tree. One issue remains: Phangorn crashes with some alignments, e.g. https://raw.githubusercontent.com/naturalis/supersmart/master/t/testdata/modeltest-aln.fa and I could not yet find out why since the alignment loks ok to me. This is therefore posted as an issue in the phangorn project: https://github.com/KlausVigo/phangorn/issues/2

**Comment by hettling** [comment]
Simulated sequences can now be inserted into the DB, also the command `smrt-utils dbinsert` works with Christine's data. 
The phangorn package is now updated on CRAN, so it does not throw an error. However, modeltest for some alignments seems to take forever. I will try to simulate a whole primate dataset on the cluster and will see if this is feasible.

**Comment by hettling** [comment]
The simulation pipeline works fine for the small quail example in `examples/`. The initially replicated tree can now be compared to the one obtained from the simulation data using `smrt-utils treedist`. Will start it now for the primates to get the data for the manuscript.

**Comment by hettling** [comment]
The simulation pipeline now creates datasets of tree/alignments that resemble an input dataset. 
SUPERSMART can then be rerun on the simulated alignments, entering at `smrt orthologize`. A workflow for simulation and subsequence tree inference can be found (here)[https://github.com/naturalis/supersmart/blob/master/examples/primates/workflow-sim.sh]. Robinson-foulds distance between replicated and re-estimated trees can than be obtained with `smrt-utils treedist`.

What do you guys think are the tools we should compare SUPERSMART to in the simulation study? Cipres has a good coverage of tools for inference, (GARLI, MrBayes, etc.) and http://www.phylogeny.fr/ offers tree inference with phyml. Running the tools on these remote servers appears convenient.

Furthermore, we have to decide what the input to these tools should be. Most tools take a single fasta/nexus/phylip file with an alignent. For more than 1000 replicated alignments for the primates, this results in a huge supermatrix, which is difficult to manage for these tools. So I'm wondering if we should 
* pick a smaller example to compare methods for species-level inference
* reduce the simulated primate dataset before re-running SUPERSMART
* compare the methods at genus-level
?

**Comment by alexantonelli** [comment]
hi guys,

I think this will need some personal discussion, on Friday’s skype?

Another thing that Hervé commented in the previous version of the manuscript is that he wasn’t confident *at all* with the analysis of the Gentianales, because of the fossil calibrations. He expected that reviewers in Syst Biol will want to see a justification of the fossil calibrations, which we don’t have since Cajsa (who selected them) hasn’t yet published the list and justifications, and will not do that in time. So he suggested that we replaced the analysis by one on the plant order Magnoliidea instead, since a comprehensive analysis was recently published by him and his PhD student, and fossils are already in the fossil calibratoin format that we made as well as in the paleontologica electronica, and he’d be able to cross-check the results. I think this would be much preferable, and if you agree we can start a thread on this including him (Hervé Sauquet <herve.sauquet@gmail.com<mailto:herve.sauquet@gmail.com>>)

Mentioned papers:
http://www.biomedcentral.com/1471-2148/15/49
http://palaeo-electronica.org/content/pdfs/435.pdf

talk to you soon!

Alex


---
Alexandre Antonelli
http://antonelli-lab.net

On 02 Sep 2015, at 17:03 , hettling <notifications@github.com<mailto:notifications@github.com>> wrote:


The simulation pipeline now creates datasets of tree/alignments that resemble an input dataset.
SUPERSMART can then be rerun on the simulated alignments, entering at smrt orthologize. A workflow for simulation and subsequence tree inference can be found (here)[https://github.com/naturalis/supersmart/blob/master/examples/primates/workflow-sim.sh]. Robinson-foulds distance between replicated and re-estimated trees can than be obtained with smrt-utils treedist.

What do you guys think are the tools we should compare SUPERSMART to in the simulation study? Cipres has a good coverage of tools for inference, (GARLI, MrBayes, etc.) and http://www.phylogeny.fr/ offers tree inference with phyml. Running the tools on these remote servers appears convenient.

Furthermore, we have to decide what the input to these tools should be. Most tools take a single fasta/nexus/phylip file with an alignent. For more than 1000 replicated alignments for the primates, this results in a huge supermatrix, which is difficult to manage for these tools. So I'm wondering if we should

  *   pick a smaller example to compare methods for species-level inference
  *   reduce the simulated primate dataset before re-running SUPERSMART
  *   compare the methods at genus-level ?

**Comment by rvosa** [comment]
Hi guys,

I guess we can discuss the exact approach for the simulation study on
friday.

As for switching to Magnoliidea instead: I am not particularly excited
about this prospect. Maybe we should do it I we have to, but the last thing
we want is for reviewers of essentially a methods paper to start
complaining about the topology of a particular example tree. Sure, the
examples shouldn't be all messed up - but the focus really needs to be on
the methods, not on whether a particular tree meshes with the beliefs a
reviewer has about the systematics of a particular group.

Rutger

**Comment by hettling** [comment]
--
Update about simulation pipeline 
--
The following issue was discovered in the simulation pipeline
* After replicating a primate tree and replicating all alignments in the tree, we constructed a backbone using the simulated data (see https://github.com/hettling/phylo-primates-sim/tree/master/results/2015-11-19-PRIMATES-NOBIN-1)
* Genera in the replicated tree were monophyletic and Families were quite well scrambled, which is the desired outcome of the tree replication
* In the backbone tree obtained from the simulated data, we saw that certain groups, such as the old-world monkeys or the hominidae were grouped together, **although they were not grouped together in the tree from which the data was simulated**! 

So there seemed to be some underlying signal grouping taxa together that do not belong together. 
First guess was that there could be **a bug in the code** (taking the wrong tree for doing the replicated alignments, taking the wrong data for re-estimating the replicated tree, etc...). Rutger and I went thoroughly through the code and the output of the simulation, but we were almost sure that the correct tree was used to replicate the alignments and the correct alignments ended up in the supermatrix.

Another possibility was that by taking the median of all sequences in an alignment as ancestral sequence to replicate the alignment along the tree, there could be some signal left.
We then tried to use a **random ancestral sequence** (commit ba32349af47c8f90c0f1a34d29ae2e5eba51de53) to simulate the alignments. Results of this workflow are in https://github.com/hettling/phylo-primates-sim/tree/master/results/2015-11-23-clusters-1.
We then replicated the orthologous cluster and not the initial alignments, since random ancestral sequences for each alignent would result in little orthologous alignments.

It looks like by starting from random ancestral sequences, we can overcome this strange grouping observed initially. 

For the simulations starting from median sequences, one can see the falso grouping very well e.g. looking at Cercopithecidae in the  [annotated replicated tree](https://github.com/hettling/phylo-primates-sim/blob/master/results/2015-11-19-PRIMATES-NOBIN-1/tree-replicated-families.dnd) and the [annotated backbone] (https://github.com/hettling/phylo-primates-sim/blob/master/results/2015-11-19-PRIMATES-NOBIN-1/consense-families.dnd).
** Robinson-foulds-distance between these trees (after pruning out terminals in the replicated tree that are not in the backbone): 65

For the simulations of the clusters from random ancestral sequences, the estimated backbone looks much better. [replicated tree](https://github.com/hettling/phylo-primates-sim/blob/master/results/2015-11-23-clusters-1/tree-replicated-families.dnd) and [backbone](https://github.com/hettling/phylo-primates-sim/blob/master/results/2015-11-23-clusters-1/consense-families.dnd). Unfortunately, in this run, the replicated tree was different than in the one above. However, trees are much more similar: ** Robinson foulds distance:  31 ** 

tl; dr (just to document what happened the last days...)

**Comment by rvosa** [comment]
Great work, thanks. How do you compute the Robinson-Foulds distance? If I recall correctly this ought to be a number between 0 and 1.

**Comment by rvosa** [comment]
I think it is _mission accomplished_ for this issue.
